File: services/scrummaster/src/registry.js
'use strict';

const fs = require('fs');
const { deriveAgentCard } = require('./a2a/agentCard');

// Canonical agent catalog (services/scrummaster/config/agents.json) and per-project
// available-agent configuration (services/scrummaster/config/projects.json).
// See the agent-assignment design REQ-01 and REQ-02.
//
// Both are loaded and validated once, at first use (effectively at process
// startup — see index.js, which calls load() eagerly so a malformed catalog
// or project config fails the process before it accepts any traffic).

let catalog = null;   // { byId: Map<string, agent>, ids: string[] }
let projects = null;  // Map<string, { name, jiraProjectKey, agents: string[] }>
let agentCardsById = null; // Map<string, AgentCard> — derived from catalog.byId

function catalogPath() {
  return process.env.AGENTS_CATALOG_PATH || '/app/config/agents.json';
}

function projectsPath() {
  return process.env.PROJECTS_CONFIG_PATH || '/app/config/projects.json';
}

function readJson(path) {
  return JSON.parse(fs.readFileSync(path, 'utf8'));
}

// Validate and index the raw agents.json contents. Throws with a message
// that names every problem found (not just the first) so a maintainer can
// fix a broken catalog in one pass.
function parseCatalog(raw, path) {
  const errors = [];
  const agents = Array.isArray(raw.agents) ? raw.agents : null;
  if (!agents) errors.push(`"agents" must be a non-empty array`);

  const seenIds = new Set();
  const byId = new Map();

  (agents || []).forEach((entry, i) => {
    const where = `agents[${i}]`;
    if (!entry || typeof entry !== 'object') {
      errors.push(`${where} must be an object`);
      return;
    }
    const { id, displayName, definitionPath, routing, agentCard } = entry;

    if (!id || typeof id !== 'string') {
      errors.push(`${where}.id is required and must be a non-empty string`);
    } else if (seenIds.has(id)) {
      errors.push(`duplicate agent id "${id}" (${where})`);
    } else {
      seenIds.add(id);
    }

    if (!displayName || typeof displayName !== 'string') {
      errors.push(`${where}.displayName is required (id: ${id || '?'})`);
    }
    if (!definitionPath || typeof definitionPath !== 'string') {
      errors.push(`${where}.definitionPath is required (id: ${id || '?'})`);
    }
    if (!routing || typeof routing !== 'object' || !routing.channelSuffix || typeof routing.channelSuffix !== 'string') {
      errors.push(`${where}.routing.channelSuffix is required (id: ${id || '?'})`);
    }
    if (!agentCard || typeof agentCard !== 'object' || !agentCard.name || !agentCard.description) {
      errors.push(`${where}.agentCard.{name,description} are required (id: ${id || '?'})`);
    }

    if (id && typeof id === 'string' && displayName && definitionPath && routing?.channelSuffix) {
      byId.set(id, entry);
    }
  });

  const retiredAgents = raw.retiredAgents === undefined ? [] : raw.retiredAgents;
  if (!Array.isArray(retiredAgents)) {
    errors.push(`"retiredAgents" must be an array if present`);
  } else {
    retiredAgents.forEach((entry, i) => {
      const where = `retiredAgents[${i}]`;
      if (!entry || typeof entry.id !== 'string' || !entry.id) {
        errors.push(`${where}.id is required`);
        return;
      }
      if (seenIds.has(entry.id)) {
        errors.push(`"${entry.id}" appears in both agents and retiredAgents (${where})`);
      }
    });
  }

  if (errors.length > 0) {
    throw new Error(`Invalid agent catalog at ${path}:\n  - ${errors.join('\n  - ')}`);
  }

  return {
    byId,
    ids: Array.from(byId.keys()),
    retiredIds: retiredAgents.map(r => r.id),
  };
}

// Validate the raw projects.json contents against the loaded catalog.
// Throws, naming the invalid project and agent id, if a project references
// an agent id that isn't a valid catalog entry.
function parseProjects(raw, path, cat) {
  const errors = [];
  const list = Array.isArray(raw.projects) ? raw.projects : null;
  if (!list) errors.push(`"projects" must be a non-empty array`);

  const byName = new Map();

  (list || []).forEach((entry, i) => {
    const where = `projects[${i}]`;
    if (!entry || typeof entry.name !== 'string' || !entry.name) {
      errors.push(`${where}.name is required`);
      return;
    }
    if (byName.has(entry.name)) {
      errors.push(`duplicate project name "${entry.name}" (${where})`);
    }
    const agentIds = Array.isArray(entry.agents) ? entry.agents : null;
    if (!agentIds) {
      errors.push(`${where}.agents must be an array (project: ${entry.name})`);
    } else {
      for (const id of agentIds) {
        if (!cat.byId.has(id)) {
          errors.push(`project "${entry.name}" references unknown agent id "${id}" — not present in agents.json`);
        }
      }
    }
    byName.set(entry.name, {
      name: entry.name,
      jiraProjectKey: entry.jiraProjectKey || null,
      agents: agentIds || [],
    });
  });

  if (errors.length > 0) {
    throw new Error(`Invalid project configuration at ${path}:\n  - ${errors.join('\n  - ')}`);
  }

  return byName;
}

function load() {
  if (catalog && projects) return;
  const cPath = catalogPath();
  const pPath = projectsPath();

  catalog = parseCatalog(readJson(cPath), cPath);
  console.log(`[registry] Loaded ${catalog.ids.length} agent(s) from ${cPath}`);

  projects = parseProjects(readJson(pPath), pPath, catalog);
  console.log(`[registry] Loaded ${projects.size} project(s) from ${pPath}`);

  // Fail fast: every registered agent must produce a valid AgentCard (REQ-08
  // of the a2a-messaging design). A catalog entry that can't
  // derive one is a configuration error, not a runtime-recoverable condition.
  agentCardsById = new Map();
  for (const id of catalog.ids) {
    agentCardsById.set(id, deriveAgentCard(catalog.byId.get(id)));
  }
}

// Look up a catalog entry by id. Returns null if the id is not a registered
// agent — callers must not treat "not found" as a fallback to any default.
function getAgent(id) {
  load();
  return catalog.byId.get(id) || null;
}

// Look up the derived AgentCard for a registered agent, or null if unknown.
function getAgentCard(id) {
  load();
  return agentCardsById.get(id) || null;
}

// All registered AgentCards, keyed by catalog agent id.
function listAgentCards() {
  load();
  return Object.fromEntries(agentCardsById);
}

// All registered agent ids, in catalog order.
function getAllAgentIds() {
  load();
  return catalog.ids.slice();
}

function getRetiredAgentIds() {
  load();
  return catalog.retiredIds.slice();
}

function getProjectNames() {
  load();
  return Array.from(projects.keys());
}

function getProject(name) {
  load();
  return projects.get(name) || null;
}

// The effective allowed-agent set for a project: catalog entries (not just
// ids) for every agent id the project's configuration enables. Returns []
// for an unconfigured project name rather than throwing — callers decide
// whether that's an error for their context.
function getEffectiveAgents(projectName) {
  load();
  const project = projects.get(projectName);
  if (!project) return [];
  return project.agents
    .map(id => catalog.byId.get(id))
    .filter(Boolean);
}

// Stream topology (the redis-streams design §4). Logical
// names are stable configuration, never accepted from an untrusted payload —
// every caller must derive `project` from the configured destination it is
// already bound to (a Jira issue's own project, or the stream a consumer is
// reading), not from message content.
const GATEWAY_GROUP = 'scrummaster';
// canonical-work-model.md REQ-15 / internal-work-item-service.md REQ-05-06:
// this service's own consumer group on the internal work-item service's
// outbound event stream (any interested subscriber creates its own group —
// see services/work-item-service/workitems/stream_topology.py's module comment).
const JIRA_CATCHUP_GROUP = 'jira-catchup';
// canonical-work-model.md REQ-21 — ScrumMaster's dispatch-trigger consumer
// group on that SAME outbound event stream. A second, independent
// subscriber (Streams' normal fan-out — every interested consumer gets its
// own group), not a replacement for JIRA_CATCHUP_GROUP.
const DISPATCH_GROUP = 'dispatch';

function normalizeProjectName(name) {
  return String(name).trim().toLowerCase();
}

// ScrumMaster -> agent: aigang:agent:{project}:{suffix}, group agent-{suffix}
function agentStreamName(projectName, suffix) {
  return `aigang:agent:${normalizeProjectName(projectName)}:${suffix}`;
}

function agentGroupName(suffix) {
  return `agent-${suffix}`;
}

// Agent -> ScrumMaster: aigang:gateway:{project}, group scrummaster
function gatewayStreamName(projectName) {
  return `aigang:gateway:${normalizeProjectName(projectName)}`;
}

// Internal work-item service -> ScrumMaster (outbound events): matches
// services/work-item-service/workitems/stream_topology.py's event_stream_name()
// exactly — the shared naming contract both sides agree on (same reasoning
// as commandStreamName in canonicalWorkItems.js).
function workItemEventStreamName(projectName) {
  return `aigang:workitems:${normalizeProjectName(projectName)}:events`;
}

module.exports = {
  load,
  // Exported for direct contract testing of catalog/project-config
  // validation without needing to write temp files or reset module state.
  parseCatalog,
  parseProjects,
  getAgent,
  getAgentCard,
  listAgentCards,
  getAllAgentIds,
  getRetiredAgentIds,
  getProjectNames,
  getProject,
  getEffectiveAgents,
  normalizeProjectName,
  agentStreamName,
  agentGroupName,
  gatewayStreamName,
  workItemEventStreamName,
  GATEWAY_GROUP,
  JIRA_CATCHUP_GROUP,
  DISPATCH_GROUP,
};

File: services/work-item-service/workitems/registry.py
"""
Catalog-backed agent/project registry — a Python reimplementation of the
parts of services/scrummaster/src/registry.js this service needs for catalog-backed
assignment validation (agent-assignment.md REQ-01/REQ-02, reused here per
canonical-work-model.md REQ-03/REQ-13's "validated against the same
catalog-backed validator agent-assignment.md already requires").

Architectural decision (see final report): a Python/Django service cannot
require() services/scrummaster/src/registry.js. Two options existed: (a) reimplement
the catalog-loading and validation logic in Python, reading the SAME
on-disk agents.json/projects.json files ScrumMaster's own registry.js
reads (AGENTS_CATALOG_PATH / PROJECTS_CONFIG_PATH — the exact same env var
names, so both services can be pointed at the same catalog files in any
deployment), or (b) call ScrumMaster's HTTP surface at runtime for every
single mutating command. (a) was chosen: the catalog is a small, rarely-
changing, already-shared-on-disk JSON configuration — not a live service
API — so reading it directly avoids adding a synchronous runtime
dependency (and a new failure mode: "work-item-service can't validate an
assignment because ScrumMaster's HTTP endpoint is down") to every write
this service accepts. The tradeoff this carries: two independent
implementations of the SAME validation rules must be kept in sync by hand
if either one changes — flagged explicitly in the final report as the
cost of this choice, not hidden.

Deliberately narrower than registry.js: this service has no A2A AgentCard
concept and no agent-catalog-authoring workflow of its own, so only the
catalog/project lookups assignment.py actually needs are ported
(getAgent, getProject, getProjectNames, normalizeProjectName). Validation
strictness (every catalog/project field required, duplicate-id detection)
is preserved because a malformed catalog should fail this service's
startup exactly as loudly as it fails ScrumMaster's.
"""

from __future__ import annotations

import json
import os
import threading
from typing import Any, Optional

from django.conf import settings

_lock = threading.Lock()
_catalog: Optional[dict[str, Any]] = None   # {'by_id': {...}, 'ids': [...], 'retired_ids': [...]}
_projects: Optional[dict[str, Any]] = None  # {name: {'name', 'jiraProjectKey', 'agents': [...]}}


def _catalog_path() -> str:
    return getattr(settings, 'AGENTS_CATALOG_PATH', None) or os.environ.get('AGENTS_CATALOG_PATH', '/app/config/agents.json')


def _projects_path() -> str:
    return getattr(settings, 'PROJECTS_CONFIG_PATH', None) or os.environ.get('PROJECTS_CONFIG_PATH', '/app/config/projects.json')


def _read_json(path: str) -> Any:
    with open(path, encoding='utf-8') as fh:
        return json.load(fh)


def _parse_catalog(raw: dict, path: str) -> dict[str, Any]:
    errors = []
    agents = raw.get('agents')
    if not isinstance(agents, list):
        errors.append('"agents" must be a non-empty array')
        agents = []

    seen_ids = set()
    by_id = {}
    for i, entry in enumerate(agents):
        where = f'agents[{i}]'
        if not isinstance(entry, dict):
            errors.append(f'{where} must be an object')
            continue
        agent_id = entry.get('id')
        display_name = entry.get('displayName')
        definition_path = entry.get('definitionPath')
        routing = entry.get('routing') or {}

        if not agent_id or not isinstance(agent_id, str):
            errors.append(f'{where}.id is required and must be a non-empty string')
        elif agent_id in seen_ids:
            errors.append(f'duplicate agent id "{agent_id}" ({where})')
        else:
            seen_ids.add(agent_id)

        if not display_name or not isinstance(display_name, str):
            errors.append(f'{where}.displayName is required (id: {agent_id or "?"})')
        if not definition_path or not isinstance(definition_path, str):
            errors.append(f'{where}.definitionPath is required (id: {agent_id or "?"})')
        if not isinstance(routing, dict) or not routing.get('channelSuffix'):
            errors.append(f'{where}.routing.channelSuffix is required (id: {agent_id or "?"})')

        if agent_id and isinstance(agent_id, str) and display_name and definition_path and routing.get('channelSuffix'):
            by_id[agent_id] = entry

    retired_agents = raw.get('retiredAgents', [])
    if not isinstance(retired_agents, list):
        errors.append('"retiredAgents" must be an array if present')
        retired_agents = []
    for i, entry in enumerate(retired_agents):
        where = f'retiredAgents[{i}]'
        if not isinstance(entry, dict) or not entry.get('id'):
            errors.append(f'{where}.id is required')
            continue
        if entry['id'] in seen_ids:
            errors.append(f'"{entry["id"]}" appears in both agents and retiredAgents ({where})')

    if errors:
        joined = '\n  - '.join(errors)
        raise ValueError(f'Invalid agent catalog at {path}:\n  - {joined}')

    return {'by_id': by_id, 'ids': list(by_id.keys()), 'retired_ids': [r['id'] for r in retired_agents]}


def _parse_projects(raw: dict, path: str, catalog: dict[str, Any]) -> dict[str, Any]:
    errors = []
    projects_list = raw.get('projects')
    if not isinstance(projects_list, list):
        errors.append('"projects" must be a non-empty array')
        projects_list = []

    by_name = {}
    for i, entry in enumerate(projects_list):
        where = f'projects[{i}]'
        name = entry.get('name') if isinstance(entry, dict) else None
        if not name or not isinstance(name, str):
            errors.append(f'{where}.name is required')
            continue
        if name in by_name:
            errors.append(f'duplicate project name "{name}" ({where})')
        agent_ids = entry.get('agents')
        if not isinstance(agent_ids, list):
            errors.append(f'{where}.agents must be an array (project: {name})')
            agent_ids = []
        else:
            for aid in agent_ids:
                if aid not in catalog['by_id']:
                    errors.append(f'project "{name}" references unknown agent id "{aid}" — not present in agents.json')
        by_name[name] = {'name': name, 'jiraProjectKey': entry.get('jiraProjectKey'), 'agents': agent_ids}

    if errors:
        joined = '\n  - '.join(errors)
        raise ValueError(f'Invalid project configuration at {path}:\n  - {joined}')

    return by_name


def load(force: bool = False) -> None:
    global _catalog, _projects
    with _lock:
        if _catalog is not None and _projects is not None and not force:
            return
        c_path = _catalog_path()
        p_path = _projects_path()
        _catalog = _parse_catalog(_read_json(c_path), c_path)
        _projects = _parse_projects(_read_json(p_path), p_path, _catalog)


def get_agent(agent_id: str) -> Optional[dict[str, Any]]:
    load()
    return _catalog['by_id'].get(agent_id)


def get_project(name: str) -> Optional[dict[str, Any]]:
    load()
    return _projects.get(name)


def get_project_names() -> list[str]:
    load()
    return list(_projects.keys())


def get_all_agent_ids() -> list[str]:
    load()
    return list(_catalog['ids'])


def normalize_project_name(name: str) -> str:
    return str(name).strip().lower()


# Stream topology names ScrumMaster's own gateway/webhook flows use
# (redis-streams.md §4) — reused verbatim so webhook_consumer.py can read
# the SAME webhookStreamName ScrumMaster's server.js already publishes to
# (see that module's comment for why: a second, independent consumer group
# on ScrumMaster's existing stream, not a new stream).
WEBHOOK_GROUP = 'scrummaster'


def webhook_stream_name(project_name: str) -> str:
    return f'aigang:webhooks:{normalize_project_name(project_name)}'
